src/clj/game/core/costs.clj

```diff
@@ -145,38 +149,151 @@
       :connection (str "trash " (quantify amount "installed connection resource"))
       :ice (str "trash " (quantify amount "installed rezzed ICE" ""))
-      (:net :meat :brain) (str "suffer " (quantify amount (str (name cost-type) " damage") ""))
-      :mill (str "trash " (quantify amount "card") " from the top of your deck")
+      ; trashing non-installed cards
+      :trash-from-deck (str "trash " (quantify amount "card") " from the top of your deck")
       :trash-from-hand (str "trash " (quantify amount "card") " from your hand")
       :randomly-trash-from-hand (str "trash " (quantify amount "card") " randomly from your hand")
+      :trash-entire-hand "trash all cards in your hand"
+      ; damage
+      (:net :meat :brain) (str "suffer " (quantify amount (str (name cost-type) " damage") ""))
+      ; counters
+      (:agenda :power :virus :advancement) (if (< 1 amount)
+                                             (quantify amount (str "hosted " (name cost-type) " counter"))
+                                             (str "hosted " (name cost-type) " counter"))
+      ; other
+      :forfeit (str "forfeit " (quantify amount "Agenda"))
+      :tag (str "remove " (quantify amount "tag"))
       :shuffle-installed-to-stack (str "shuffle " (quantify amount "installed card") " into the stack")
-      (str "pay " (quantify amount (name cost-type))))))
+      (str (quantify amount (name cost-type))))))
 
-(defn build-cost-str
+(defn build-cost-label
   "Gets the complete cost-str for specified costs"
   [costs]
   (let [cost-string
         (->> costs
              merge-costs
-             (map #(cost-names %))
+             (map cost->label)
              (filter some?)
-             (interpose " and ")
+             (interpose ", ")
              (apply str))]
-    (capitalize cost-string)))
+    (when (not (string/blank? cost-string))
+      (capitalize cost-string))))
+
+(defn make-label
+  "Looks into an ability for :label, if it doesn't find it, capitalizes :msg instead."
+  [ability]
+  (let [label (or (:label ability)
+                  (and (string? (:msg ability))
+                       (capitalize (:msg ability)))
+                  "")
+        cost (concat (:cost ability)
+                     (when-let [cost (:counter-cost ability)]
+                       (when (not= :credit (first cost))
+                         cost))
+                     (when-let [cost (:advance-counter-cost ability)]
+                       [:advancement cost]))]
+    (if (and (seq cost) (not (string/blank? label)))
+      (str (build-cost-label cost) ": " label)
+      label)))
+
+(let [cost-types #{:trash
+                   :hardware :program :resource :connection :ice
+                   :net :meat :brain
+                   :trash-from-deck :trash-from-hand :randomly-trash-from-hand
+                   :agenda :power :virus :advancement
+                   :forfeit :shuffle-installed-to-stack}]
+  (defn cost->string
```

Comment:
Maybe swap `defn` and `let`?

Reply:
I wrote it like this because I don't want to recreate the `cost-types` set every call to `cost->string`, as it's static, but I also don't want to `def` it into a namespace-global variable, because that's messy. Doing the `let` binding around the `defn` means that it'll only be created once, and only this function will have access to it.

Reply:
Decided to reverse this logic, so it's not even needed anymore!